src/NuGet.Services.AzureSearch/Catalog2AzureSearch/CatalogLeafFetcher.cs

```diff
@@ -0,0 +1,253 @@
+                }
+
+                if (!info.VersionToItem.TryGetValue(version, out var item))
+                {
+                    _logger.LogWarning(
+                        "No registration leaf item found for {PackageId} {Version} on {PageUrl}",
+                        packageId,
+                        version,
+                        info.Page.Url);
+                    unavailable.Add(version);
+                    continue;
+                }
+
+                _logger.LogInformation(
+                    "Fetching the catalog leaf for {PackageId} {Version} from {LeafUrl}",
+                    packageId,
+                    version,
+                    item.CatalogEntry.Url);
+
+                var leaf = await _catalogClient.GetPackageDetailsLeafAsync(item.CatalogEntry.Url);
+                available[version] = leaf;
+
+                if (leaf.IsListed())
+                {
+                    _logger.LogInformation(
+                        "{PackageId} {Version} was found to be listed. Metadata from {Url} will be used.",
+                        packageId,
+                        version,
+                        leaf.Url);
+                    return;
+                }
+                else
+                {
+                    _logger.LogInformation(
+                        "{PackageId} {Version} was found to be unlisted. This will not be used as a latest version.",
+                        packageId,
+                        version,
+                        leaf.Url);
+                }
+            }
+
+            _logger.LogWarning(
+                "No catalog leaves matchings leaves found for {PackageId}. Versions tried: {Versions}",
+                packageId,
+                descendingVersions);
+        }
+
+        private string GetRegistrationIndexUrl(string id)
+        {
+            var baseUrl = _options.Value.RegistrationsBaseUrl.TrimEnd('/');
+            return $"{baseUrl}/{id.ToLowerInvariant()}/index.json";
+        }
+
+        private RegistrationPageInfo GetPageInfo(
+            IReadOnlyDictionary<string, RegistrationPageInfo> pageUrlToInfo,
+            NuGetVersion version)
+        {
+            foreach (var info in pageUrlToInfo.Values)
+            {
+                if (info.Range.Satisfies(version))
+                {
+                    return info;
+                }
+            }
+
+            return null;
+        }
+
+        private class RegistrationPageInfo
+        {
+            public RegistrationPageInfo(RegistrationPage page)
+            {
+                Page = page;
+                Range = new VersionRange(
+                    minVersion: NuGetVersion.Parse(page.Lower),
+                    includeMinVersion: true,
+                    maxVersion: NuGetVersion.Parse(page.Upper),
+                    includeMaxVersion: true);
+                RangeString = Range.ToNormalizedString();
+
+                if (page.Items != null)
+                {
```

Comment:
Is this null check necessary? Are there registration pages with null items? The docs say `items` is a required property: [https://docs.microsoft.com/en-us/nuget/api/registration-base-url-resource#registration-page](https://docs.microsoft.com/en-us/nuget/api/registration-base-url-resource#registration-page)

I'm only asking this out of curiosity, it's definitely better to be defensive around null values :)